modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/ForEachProcessor.java

```diff
@@ -67,20 +67,61 @@ boolean isIgnoreMissing() {
 
     @Override
     public void execute(IngestDocument ingestDocument, BiConsumer<IngestDocument, Exception> handler) {
-        List<?> values = ingestDocument.getFieldValue(field, List.class, ignoreMissing);
-        if (values == null) {
+        Object o = ingestDocument.getFieldValue(field, Object.class, ignoreMissing);
+        if (o == null) {
             if (ignoreMissing) {
                 handler.accept(ingestDocument, null);
             } else {
                 handler.accept(null, new IllegalArgumentException("field [" + field + "] is null, cannot loop over its elements."));
             }
+        } else if (o instanceof Map) {
+            @SuppressWarnings("unchecked")
+            Map<String, ?> map = (Map<String, ?>) o;
```

Comment:
I think we can cast to `Map<?, ?>` instead? Then the SuppressWarnings can be removed. In order to get the key a cast to String would required, but that shouldn't generate warnings.